File: mobile/app/src/main/java/com/esousa/beam/blob/BlobSender.kt
package com.esousa.beam.blob

import com.esousa.beam.protocol.BlobBeginPayload
import com.esousa.beam.protocol.BlobEndPayload
import com.esousa.beam.protocol.Envelope
import com.esousa.beam.protocol.MessageType
import com.esousa.beam.protocol.Protocol
import com.esousa.beam.protocol.SecureEnvelope
import com.esousa.beam.security.SessionCipher
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import okhttp3.WebSocket

object BlobSender {
    fun send(
        socket: WebSocket,
        cipher: SessionCipher,
        json: Json,
        data: ByteArray,
        metadataFactory: (String) -> Envelope,
    ): String {
        val blobIdBytes = BlobChunkCodec.createBlobIdBytes()
        val blobId = blobIdBytes.toHex()
        val begin = Envelope(
            type = MessageType.BLOB_BEGIN,
            id = newId(),
            timestamp = System.currentTimeMillis(),
            payload = json.encodeToJsonElement(
                BlobBeginPayload(
                    blobId = blobId,
                    totalBytes = data.size.toLong(),
                    chunkSize = Protocol.DEFAULT_CHUNK_SIZE,
                    sha256 = BlobChunkCodec.sha256Hex(data),
                ),
            ).jsonObject,
        )
        sendEnvelope(socket, SecureEnvelope.encrypt(begin, cipher, json), json)

        var sequence = 0
        var offset = 0
        while (offset < data.size) {
            val length = minOf(Protocol.DEFAULT_CHUNK_SIZE, data.size - offset)
            val frame = BlobChunkCodec.encodeChunk(
                blobIdBytes,
                sequence,
                data.copyOfRange(offset, offset + length),
                cipher,
            )
            check(socket.send(okio.ByteString.of(*frame))) { "Falha ao enviar chunk." }
            offset += length
            sequence++
        }

        val end = Envelope(
            type = MessageType.BLOB_END,
            id = newId(),
            timestamp = System.currentTimeMillis(),
            payload = json.encodeToJsonElement(BlobEndPayload(blobId)).jsonObject,
        )
        sendEnvelope(socket, SecureEnvelope.encrypt(end, cipher, json), json)
        sendEnvelope(socket, SecureEnvelope.encrypt(metadataFactory(blobId), cipher, json), json)
        return blobId
    }

    private fun sendEnvelope(socket: WebSocket, envelope: Envelope, json: Json) {
        check(socket.send(json.encodeToString(Envelope.serializer(), envelope))) {
            "Falha ao enviar envelope."
        }
    }

    private fun newId(): String = java.util.UUID.randomUUID().toString().replace("-", "")

    private fun ByteArray.toHex(): String = joinToString(separator = "") { byte ->
        "%02x".format(byte)
    }
}

File: desktop/src/Beam.Core/Net/BlobSender.cs
using System.Net.WebSockets;
using Beam.Core.Protocol;
using Beam.Core.Security;

namespace Beam.Core.Net;

/// <summary>Envia um blob completo (begin → chunks → end) por WebSocket.</summary>
public static class BlobSender
{
    public static async Task<string> SendAsync(
        WebSocket socket,
        SessionCipher cipher,
        byte[] data,
        Func<Envelope, Envelope> encrypt,
        CancellationToken ct)
    {
        var blobIdBytes = BlobChunkCodec.CreateBlobIdBytes();
        var blobId = BlobChunkCodec.ToBlobIdString(blobIdBytes);
        var sha256 = BlobChunkCodec.ComputeSha256Hex(data);
        var chunkSize = ProtocolInfo.DefaultChunkSize;

        await SendEnvelopeAsync(
            socket,
            encrypt(Envelope.Create(MessageType.BlobBegin, new BlobBeginPayload(blobId, data.Length, chunkSize, sha256))),
            ct).ConfigureAwait(false);

        var sequence = 0u;
        for (var offset = 0; offset < data.Length; offset += chunkSize, sequence++)
        {
            var length = Math.Min(chunkSize, data.Length - offset);
            var frame = BlobChunkCodec.EncodeChunk(blobIdBytes, sequence, data.AsSpan(offset, length), cipher);
            await socket.SendAsync(frame, WebSocketMessageType.Binary, true, ct).ConfigureAwait(false);
        }

        await SendEnvelopeAsync(
            socket,
            encrypt(Envelope.Create(MessageType.BlobEnd, new BlobEndPayload(blobId))),
            ct).ConfigureAwait(false);

        return blobId;
    }

    private static Task SendEnvelopeAsync(WebSocket socket, Envelope envelope, CancellationToken ct) =>
        socket.SendAsync(
            System.Text.Encoding.UTF8.GetBytes(MessageSerializer.Serialize(envelope)),
            WebSocketMessageType.Text,
            true,
            ct);
}
